Support motion photos

kvalev opened the pull request:
related to https://github.com/photoprism/photoprism/issues/439

**caveats**:
- loading the video for a Google/Samung motion photo for the very first time (on hover) can take few seconds, as the video has to be transcoded on demand
- ~~the feature only works if exiftool is enabled (as it relies on the improved metadata extraction)~~

sources:
- [MotionPhoto / MicroVideo File Formats on Pixel Phones](https://timojyrinki.gitlab.io/hugo/post/2021-03-30-pixel-motionphoto-microvideo-file-formats/)
- [Working with Motion Photos](https://medium.com/android-news/working-with-motion-photos-da0aa49b50c)
- [How to extract the Photo/Video component of a MVIMG?](https://stackoverflow.com/questions/53104989/how-to-extract-the-photo-video-component-of-a-mvimg)

kvalev committed:
MP: Parse additional exif metadata

kvalev committed:
MP: Extract embedded video and show as live photo

kvalev committed:
MP: Support legacy Google motion photos format

kvalev committed:
MP: Support Samsung motion photos

kvalev committed:
MP: Reduce reliance on exiftool

kvalev committed:
MP: Improve error messages and documentation

rickysarraf commented:
Hi @kvalev 

I hope it is okay to give you some feedback on this MR. I've been waiting for **Motion Photos** support.

While indexing, I get the following crash:

```
photoprism_1  | time="2021-09-14T16:47:53Z" level=info msg="index: extracted 2 faces from MVIMG_20201007_170906.mp4.jpg [2.947178379s]"
photoprism_1  | time="2021-09-14T16:47:54Z" level=info msg="media: 11 thumbs created for MVIMG_20201007_170930 [1.510657279s]"
photoprism_1  | time="2021-09-14T16:47:54Z" level=info msg="index: added related mp4 file MVIMG_20201007_170930.mp4"
photoprism_1  | time="2021-09-14T16:47:54Z" level=info msg="media: MVIMG_20201007_170930.mp4.jpg was taken at 2020-10-07 00:00:00 +0000 UTC (name)"
photoprism_1  | time="2021-09-14T16:47:54Z" level=info msg="index: added related jpg file MVIMG_20201007_170906.mp4.jpg"
photoprism_1  | time="2021-09-14T16:47:54Z" level=info msg="index: ignored \"Digikam_TO_BE_FILTERED/MVIMG_20201013_110639.xmp\""
photoprism_1  | time="2021-09-14T16:47:54Z" level=info msg="mp: extracting mp4 from motion photo Digikam_TO_BE_FILTERED/MVIMG_20201013_110639.jpg"
photoprism_1  | time="2021-09-14T16:47:54Z" level=info msg="mp: detected that Digikam_TO_BE_FILTERED/MVIMG_20201013_110639.jpg is a Google legacy mo
tion photo"
photoprism_1  | panic: runtime error: slice bounds out of range [-449898:] 
photoprism_1  | 
photoprism_1  | goroutine 537 [running]:
photoprism_1  | github.com/photoprism/photoprism/internal/photoprism.(*MediaFile).ExtractVideoFromMotionPhoto(0xc003f7d880)
photoprism_1  |         /go/src/github.com/photoprism/photoprism/internal/photoprism/mediafile_motionphoto.go:100 +0xe45
photoprism_1  | github.com/photoprism/photoprism/internal/photoprism.IndexMain(0xc0042c3ed8, 0xc00160a800, {{0x1fda2d8, 0x1}, 0x1, 0x1, 0x1})
photoprism_1  |         /go/src/github.com/photoprism/photoprism/internal/photoprism/index_related.go:41 +0x39a
photoprism_1  | github.com/photoprism/photoprism/internal/photoprism.IndexRelated({{0xc002ecca20, 0x2, 0x2}, 0xc003f7d880}, 0xc00160a800, {{0x1fda2d
8, 0x1}, 0x1, 0x1, 0x1})
photoprism_1  |         /go/src/github.com/photoprism/photoprism/internal/photoprism/index_related.go:88 +0x145
photoprism_1  | github.com/photoprism/photoprism/internal/photoprism.IndexWorker(0xc0070740a0)
photoprism_1  |         /go/src/github.com/photoprism/photoprism/internal/photoprism/index_worker.go:12 +0x108
photoprism_1  | github.com/photoprism/photoprism/internal/photoprism.(*Index).Start.func2()
photoprism_1  |         /go/src/github.com/photoprism/photoprism/internal/photoprism/index.go:102 +0x26
photoprism_1  | created by github.com/photoprism/photoprism/internal/photoprism.(*Index).Start
photoprism_1  |         /go/src/github.com/photoprism/photoprism/internal/photoprism/index.go:101 +0x3cd
mariadb_1     | 2021-09-14 16:47:55 43 [Warning] Aborted connection 43 to db: 'photoprism' user: 'photoprism' host: '172.17.0.3' (Got an error readi
ng communication packets)
mariadb_1     | 2021-09-14 16:47:55 42 [Warning] Aborted connection 42 to db: 'photoprism' user: 'photoprism' host: '172.17.0.3' (Got an error readi
ng communication packets)
mariadb_1     | 2021-09-14 16:47:55 41 [Warning] Aborted connection 41 to db: 'photoprism' user: 'photoprism' host: '172.17.0.3' (Got an error readi
ng communication packets)
mariadb_1     | 2021-09-14 16:47:55 40 [Warning] Aborted connection 40 to db: 'photoprism' user: 'photoprism' host: '172.17.0.3' (Got an error read$
```

kvalev commented:
hey @rickysarraf, thanks a lot for the feedback, I really appreciate it!

Do you mind running exiftool and pasting the output (I only need the `[XMP]` tags):
```
exiftool -U -G -a Digikam_TO_BE_FILTERED/MVIMG_20201013_110639.jpg
```

EDIT: And also the file size in bytes would be helpful

rickysarraf commented:
> hey @rickysarraf, thanks a lot for the feedback, I really appreciate it!
> 
> Do you mind running exiftool and pasting the output (I only need the `[XMP]` tags):
> 
> ```
> exiftool -U -G -a Digikam_TO_BE_FILTERED/MVIMG_20201013_110639.jpg
> ```
> 

Here it is @kvalev 

```
[EXIF]          Resolution Unit                 : inches
[XMP]           XMP Toolkit                     : XMP Core 4.4.0-Exiv2
[XMP]           Micro Video                     : 1
[XMP]           Micro Video Version             : 1
[XMP]           Micro Video Offset              : 1528953
[XMP]           Micro Video Presentation Timestamp Us: 0
[XMP]           Pick Label                      : 0
[XMP]           Color Label                     : 0
[XMP]           Date/Time Original              : 2020:10:13 11:06:39
[XMP]           Exif Version                    : 0231
[XMP]           Flashpix Version                : 0100
[XMP]           Color Space                     : Uncalibrated
[XMP]           Exif Image Width                : 3000
[XMP]           Exif Image Height               : 4000
[XMP]           Exposure Time                   : 1/33
[XMP]           F Number                        : 1.8
[XMP]           Exposure Program                : Program AE
[XMP]           Shutter Speed Value             : 1/33
[XMP]           Aperture Value                  : 1.8
[XMP]           Brightness Value                : 1.89
[XMP]           Exposure Compensation           : 0
[XMP]           Max Aperture Value              : 1.8
[XMP]           Subject Distance                : 0.68 m
[XMP]           Metering Mode                   : Center-weighted average
[XMP]           Focal Length                    : 4.7 mm
[XMP]           Sensing Method                  : One-chip color area
[XMP]           Scene Type                      : Directly photographed
[XMP]           Custom Rendered                 : Custom
[XMP]           Exposure Mode                   : Auto
[XMP]           White Balance                   : Auto
[XMP]           Digital Zoom Ratio              : 0
[XMP]           Focal Length In 35mm Format     : 26 mm
[XMP]           Scene Capture Type              : Standard
[XMP]           Contrast                        : Normal
[XMP]           Saturation                      : Normal
[XMP]           Sharpness                       : Normal
[XMP]           Subject Distance Range          : Macro
[XMP]           GPS Version ID                  : 2.0.0.0
[XMP]           GPS Map Datum                   : WGS-84
[XMP]           GPS Altitude Ref                : Above Sea Level
[XMP]           GPS Altitude                    : 0 m
[XMP]           GPS Latitude Ref                : N
[XMP]           GPS Latitude                    : 12 deg 54' 35.50" N
[XMP]           GPS Longitude Ref               : E
[XMP]           GPS Longitude                   : 77 deg 52' 39.84" E
[XMP]           Date/Time Modified              : 2020:10:13 11:06:39
[XMP]           Image Width                     : 3000
[XMP]           Image Height                    : 4000
[XMP]           Orientation                     : Horizontal (normal)
[XMP]           Y Cb Cr Positioning             : Centered
[XMP]           X Resolution                    : 72
[XMP]           Y Resolution                    : 72
[XMP]           Resolution Unit                 : inches
[XMP]           Make                            : google
[XMP]           Camera Model Name               : Pixel 4 XL
[XMP]           Software                        : Google
[XMP]           Metadata Date                   : 2020:10:13 11:06:39
[XMP]           Create Date                     : 2020:10:13 11:06:39
[XMP]           Modify Date                     : 2020:10:13 11:06:39
[XMP]           Date Time Original              : 2020:10:13 11:06:39
[XMP]           Date UTC                        : 2020:10:13 11:06:39
[XMP]           Modification Date               : 2020:10:13 11:06:39
[XMP]           Date Created                    : 2020:10:13 11:06:39
[XMP]           Urgency                         : 0 (reserved)
[XMP]           Categories                      : <Categories><Category Assigned="0">People<Category Assigned="0">Family<Category Assigned="1">Anshi
ka Raj Sarraf</Category><Category Assigned="1">Ritesh Raj Sarraf</Category></Category></Category></Categories>
[XMP]           Tags List                       : People/Family/Ritesh Raj Sarraf, People/Family/Anshika Raj Sarraf
[XMP]           Flash Fired                     : False
[XMP]           Flash Return                    : No return detection
[XMP]           Flash Mode                      : Auto
[XMP]           Flash Function                  : False
[XMP]           Flash Red Eye Mode              : False
[XMP]           Components Configuration        : Y, Cb, Cr, -
[XMP]           ISO                             : 90
[XMP]           Last Keyword XMP                : People/Family/Ritesh Raj Sarraf, People/Family/Anshika Raj Sarraf
[XMP]           Hierarchical Subject            : People|Family|Ritesh Raj Sarraf, People|Family|Anshika Raj Sarraf
[XMP]           Catalog Sets                    : People|Family|Ritesh Raj Sarraf, People|Family|Anshika Raj Sarraf
[XMP]           Region Name                     : Anshika Raj Sarraf, Ritesh Raj Sarraf
[XMP]           Region Type                     : Face, Face
[XMP]           Region Area X                   : 0.791167, 0.331
[XMP]           Region Area Y                   : 0.365, 0.17175
[XMP]           Region Area W                   : 0.195667, 0.253333
[XMP]           Region Area H                   : 0.19, 0.27
[XMP]           Region Area Unit                : normalized, normalized
[XMP]           Region Person Display Name      : Anshika Raj Sarraf, Ritesh Raj Sarraf
[XMP]           Region Rectangle                : 0.693333, 0.27, 0.195667, 0.19, 0.204333, 0.03675, 0.253333, 0.27
[XMP]           Subject                         : Ritesh Raj Sarraf, Anshika Raj Sarraf
```
> EDIT: And also the file size in bytes would be helpful

```
$ ls -l MVIMG_20201013_110639.jpg 
-rwxrwx--- 1 rrs rrs 1079055 Jul 30 23:45 MVIMG_20201013_110639.jpg*
```


The pictures were not taken with a Google Pixel phone.
They were instead taken with the ported gcam software for my phone (Xiaomi Redmi Note)

kvalev committed:
MP: Prevent overflow in case of invalid metadata

kvalev commented:
Given those numbers the error makes sense - according to the metadata, the motion photo should be 1528953 bytes, while the whole file is 1079055 bytes. So something is probably wrong with the photo.

I have added a new commit that will catch such errors and will log them (instead of crashing). Can you please try it out with a bunch more photos from your library and let me know how it went. I am not sure whether only this motion photo is corrupt or whether all/most of them have the same issue. It might also be the case that the GCam port writes some funny numbers to the metadata, instead of the proper offset.

Have you tried visualizing this specific photo in a different software, e.g. Google Photos? Does it show the motion photo correctly?

rickysarraf commented:
> Given those numbers the error makes sense - according to the metadata, the motion photo should be 1528953 bytes, while the whole file is 1079055 bytes. So something is probably wrong with the photo.
> 

Yes. That file does seem to have some odd numbers. For example, under Google Photos, it says that the size is 768 KiB while its a 12 Megapixel photograph. 

> I have added a new commit that will catch such errors and will log them (instead of crashing). Can you please try it out with a bunch more photos from your library and let me know how it went. I am not sure whether only this motion photo is corrupt or whether all/most of them have the same issue. It might also be the case that the GCam port writes some funny numbers to the metadata, instead of the proper offset.

Yes. I have triggered a rescan of the library and I'll share my results with you. Nevertheless I agree that irrespective of the picture type, the software shouldn't crash.

> 
> Have you tried visualizing this specific photo in a different software, e.g. Google Photos? Does it show the motion photo correctly?

Surprisingly, Yes. The motion effects are very small duration in this photo, like just 1.5 seconds. But GPhotos is able to render it well.

kvalev commented:
> > Have you tried visualizing this specific photo in a different software, e.g. Google Photos? Does it show the motion photo correctly?
> 
> Surprisingly, Yes. The motion effects are very small duration in this photo, like just 1.5 seconds. But GPhotos is able to render it well.

This is very interesting. I am curious how they handle that. If it is okay for you from privacy standpoint, can you send me the photo in question (especially if it is the only one failing)? I will only use it to try and debug the problem locally. You can see my email in the Git history.

rickysarraf commented:
> This is very interesting. I am curious how they handle that. If it is okay for you from privacy standpoint, can you send me the photo in question (especially if it is the only one failing)? I will only use it to try and debug the problem locally. You can see my email in the Git history.

I've emailed you the motion photo.

rickysarraf commented:
I've finally got to test this feature with some of the Motion Photos that have been scanned. With this PR, the *Support for Motion Photos* is so well integrated. Thank you. 🙏🏽

kvalev commented:
> I've finally got to test this feature with some of the Motion Photos that have been scanned. With this PR, the _Support for Motion Photos_ is so well integrated. Thank you. 🙏🏽

Happy to hear 😄 Did you have issues with any other photos?

I brute forced the photo you sent me and it does indeed contain a motion photo, but the offset in the metadata is completely wrong. I am not sure how good of an idea it is to search the whole file for a random mp4 (especially when there can be false positives). I will check whether there is a better method than the one I am using and decide whether it makes sense to incorporate it.

kvalev committed:
MP: Fallback in case of erroneous offset metadata

rickysarraf commented:
> > I've finally got to test this feature with some of the Motion Photos that have been scanned. With this PR, the _Support for Motion Photos_ is so well integrated. Thank you. 🙏🏽
> 
> Happy to hear smile Did you have issues with any other photos?
> 

No. So far, this has been the only crash.

> I brute forced the photo you sent me and it does indeed contain a motion photo, but the offset in the metadata is completely wrong. I am not sure how good of an idea it is to search the whole file for a random mp4 (especially when there can be false positives). I will check whether there is a better method than the one I am using and decide whether it makes sense to incorporate it.

I have other *Motion Photos* which do not render in PhotoPrism. But that possibly is because of the same reason as above. But that is fine, we can't cover all the cases especially given that there'll be many many users who use the gcam port. The important bit is that PhotoPrism shouldn't crash on broken photos, which seems to be fixed now.

kvalev commented:
The latest commit adds a fallback for this use case, so I expect this will fix the other motion photos as well. If you want you can give it a go and let me know how it went.

kvalev committed:
MP: Read Samsung trailer metadata

kvalev committed:
MP: Unit tests to check if a media is a motion photo

rickysarraf commented:
> The latest commit adds a fallback for this use case, so I expect this will fix the other motion photos as well. If you want you can give it a go and let me know how it went.

I tried the latest changes over the weekend. Ran a full rescan. There are a bunch of (b0rken) Motion Photos which don't show up under PhotoPrism, but are somehow working under GPhotos.

The error I get for the faulty MP, when hovering over the Photo, is:

```
photoprism_1  | time="2021-09-20T05:25:58Z" level=info msg="photos: found 60 results for public:true dist:20 quality:3 count:60 offset:120 order:newest merged:true [260.40168ms]"
photoprism_1  | time="2021-09-20T05:26:04Z" level=error msg="video: record not found"
photoprism_1  | time="2021-09-20T05:26:13Z" level=error msg="video: record not found"
```

kvalev commented:
> I tried the latest changes over the weekend. Ran a full rescan. There are a bunch of (b0rken) Motion Photos which don't show up under PhotoPrism, but are somehow working under GPhotos.

That's very interesting. Do you mind sending me one or two of those photos, so I can take a look?

kvalev committed:
MP: Fallback for Samsung motion photos in case exiftool is disabled

rickysarraf commented:
> 
> That's very interesting. Do you mind sending me one or two of those photos, so I can take a look?

I've privately emailed you the link for some sample pictures.

rickysarraf commented:
I managed to produce another crash today. This is with the latest `preview` branch from upstream, merged with yours'.

```
Sep 24 22:17:45 lenovo systemd[4444]: Starting Trigger PhotoPrism Index update...
Sep 24 22:17:51 lenovo docker[3053254]: time="2021-09-24T16:47:51Z" level=info msg="indexing originals in /photoprism/originals"
Sep 24 22:17:51 lenovo docker[3053254]: time="2021-09-24T16:47:51Z" level=info msg="classify: loading nasnet"
Sep 24 22:17:55 lenovo docker[3053254]: time="2021-09-24T16:47:55Z" level=info msg="classify: loading labels from labels.txt"
Sep 24 22:17:55 lenovo docker[3053254]: time="2021-09-24T16:47:55Z" level=info msg="converting Digikam_TO_BE_FILTERED/PXL_20210308_122423494~2.mp4.1615249859934 to jpg"
Sep 24 22:17:55 lenovo docker[3053254]: time="2021-09-24T16:47:55Z" level=warning msg="index: failed converting PXL_20210308_122423494~2.mp4.1615249859934 to jpeg (ffmpeg version 4.3.2-0+deb11u1ubuntu1 Copyright (c) 2000-2021 the FFmpeg developers\n  built with gcc 10 (Ubuntu 10.2.1-20ubuntu1)\n  configuration: --prefix=/usr --extra-version=0+deb11u1ubuntu1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-avresample --disable-filter=resample --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-pocketsphinx --enable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-nvenc --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared\n  WARNING: library configuration mismatch\n  avcodec     configuration: --prefix=/usr --extra-version=0+deb11u1ubuntu1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-avresample --disable-filter=resample --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librsvg --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwavpack --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-pocketsphinx --enable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-nvenc --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared --enable-version3 --disable-doc --disable-programs --enable-libaribb24 --enable-liblensfun --enable-libopencore_amrnb --enable-libopencore_amrwb --enable-libtesseract --enable-libvo_amrwbenc\n  libavutil      56. 51.100 / 56. 51.100\n  libavcodec     58. 91.100 / 58. 91.100\n  libavformat    58. 45.100 / 58. 45.100\n  libavdevice    58. 10.100 / 58. 10.100\n  libavfilter     7. 85.100 /  7. 85.100\n  libavresample   4.  0.  0 /  4.  0.  0\n  libswscale      5.  7.100 /  5.  7.100\n  libswresample   3.  7.100 /  3.  7.100\n  libpostproc    55.  7.100 / 55.  7.100\n[mov,mp4,m4a,3gp,3g2,mj2 @ 0x55c65019d380] moov atom not found\n/photoprism/originals/Digikam_TO_BE_FILTERED/PXL_20210308_122423494~2.mp4.1615249859934: Invalid data found when processing input\n)"
Sep 24 22:17:59 lenovo docker[3053254]: time="2021-09-24T16:47:59Z" level=info msg="index: ignored \"Pictures/NTAP/Netapp_Hyd/Thumbs.db\""
Sep 24 22:18:00 lenovo docker[3053254]: time="2021-09-24T16:48:00Z" level=info msg="index: ignored \"Pictures/Nostalgia/Thumbs.db\""
Sep 24 22:18:00 lenovo docker[3053254]: time="2021-09-24T16:48:00Z" level=info msg="index: ignored \"Pictures/Thumbs.db\""
Sep 24 22:18:00 lenovo docker[3053254]: time="2021-09-24T16:48:00Z" level=info msg="index: ignored \"Pictures/Tour De Ooty/'06_01_09_01/DCIM/101MSDCF/Thumbs.db\""
Sep 24 22:18:01 lenovo docker[3053254]: time="2021-09-24T16:48:01Z" level=info msg="index: ignored \"Pictures/Tour De Ooty/Amresh/Thumbs.db\""
Sep 24 22:18:01 lenovo docker[3053254]: time="2021-09-24T16:48:01Z" level=info msg="index: ignored \"Pictures/Tour De Ooty/Ravi's Camera/Thumbs.db\""
Sep 24 22:18:01 lenovo docker[3053254]: time="2021-09-24T16:48:01Z" level=info msg="index: ignored \"Pictures/digikam4.db\""
Sep 24 22:18:01 lenovo docker[3053254]: time="2021-09-24T16:48:01Z" level=info msg="index: ignored \"Pictures/exif-info.db\""
Sep 24 22:18:01 lenovo docker[3053254]: time="2021-09-24T16:48:01Z" level=info msg="index: ignored \"Pictures/recognition.db\""
Sep 24 22:18:01 lenovo docker[3053254]: time="2021-09-24T16:48:01Z" level=info msg="index: ignored \"Pictures/similarity.db\""
Sep 24 22:18:01 lenovo docker[3053254]: time="2021-09-24T16:48:01Z" level=info msg="index: ignored \"Pictures/thumbnails-digikam.db\""
Sep 24 22:18:02 lenovo docker[3053254]: time="2021-09-24T16:48:02Z" level=info msg="mp: extracting mp4 from motion photo Processed/photos-backup-rmi9/PXL_20210414_155853538.MP.jpg"
Sep 24 22:18:02 lenovo docker[3053254]: time="2021-09-24T16:48:02Z" level=info msg="mp: detected that Processed/photos-backup-rmi9/PXL_20210414_155853538.MP.jpg is a Google motion photo"
Sep 24 22:18:02 lenovo docker[3053254]: panic: runtime error: slice bounds out of range [-2557844:]
Sep 24 22:18:02 lenovo docker[3053254]: goroutine 70 [running]:
Sep 24 22:18:02 lenovo docker[3053254]: github.com/photoprism/photoprism/internal/photoprism.(*MediaFile).ExtractVideoFromMotionPhoto(0xc00439c380)
Sep 24 22:18:02 lenovo docker[3053254]:         /go/src/github.com/photoprism/photoprism/internal/photoprism/mediafile_motionphoto.go:88 +0x1265
Sep 24 22:18:02 lenovo docker[3053254]: github.com/photoprism/photoprism/internal/photoprism.IndexMain(0xc001c09ed8, 0xc00169fd00, {{0x0, 0x0}, 0x0, 0x1, 0x1, 0x0})
Sep 24 22:18:02 lenovo docker[3053254]:         /go/src/github.com/photoprism/photoprism/internal/photoprism/index_related.go:41 +0x34a
Sep 24 22:18:02 lenovo docker[3053254]: github.com/photoprism/photoprism/internal/photoprism.IndexRelated({{0xc003c7b5c0, 0x1, 0x1}, 0xc00439c380}, 0xc00169fd00, {{0x0, 0x0}, 0x0, 0x1, 0x1, ...})
Sep 24 22:18:02 lenovo docker[3053254]:         /go/src/github.com/photoprism/photoprism/internal/photoprism/index_related.go:88 +0x14a
Sep 24 22:18:02 lenovo docker[3053254]: github.com/photoprism/photoprism/internal/photoprism.IndexWorker(0x0)
Sep 24 22:18:02 lenovo docker[3053254]:         /go/src/github.com/photoprism/photoprism/internal/photoprism/index_worker.go:12 +0xe5
Sep 24 22:18:02 lenovo docker[3053254]: github.com/photoprism/photoprism/internal/photoprism.(*Index).Start.func2()
Sep 24 22:18:02 lenovo docker[3053254]:         /go/src/github.com/photoprism/photoprism/internal/photoprism/index.go:117 +0x26
Sep 24 22:18:02 lenovo docker[3053254]: created by github.com/photoprism/photoprism/internal/photoprism.(*Index).Start
Sep 24 22:18:02 lenovo docker[3053254]:         /go/src/github.com/photoprism/photoprism/internal/photoprism/index.go:116 +0x41e
Sep 24 22:18:02 lenovo systemd[4444]: photoprism-index.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Sep 24 22:18:02 lenovo systemd[4444]: photoprism-index.service: Failed with result 'exit-code'.
Sep 24 22:18:02 lenovo systemd[4444]: Failed to start Trigger PhotoPrism Index update.
```


The XMP tags are:

```
[XMP]           XMP Toolkit                     : XMP Core 4.4.0-Exiv2                                                                                                                                                                                                                     [18/16623]
[XMP]           Motion Photo                    : 1        
[XMP]           Motion Photo Presentation Timestamp Us: 551276
[XMP]           Motion Photo Version            : 1      
[XMP]           Pick Label                      : 0
[XMP]           Color Label                     : 0                                                                                                                                                                                                                                                  
[XMP]           Date/Time Original              : 2021:04:14 15:58:53                                                                                                                                                                                                                                
[XMP]           Exif Version                    : 0231                                                                                                                                                                                                                                               
[XMP]           Flashpix Version                : 0100                                                                                            
[XMP]           Color Space                     : sRGB            
[XMP]           Exif Image Width                : 2304                                                                                            
[XMP]           Exif Image Height               : 1728                                                                                            
[XMP]           Exposure Time                   : 1/20                                                                                            
[XMP]           F Number                        : 2.2                                                                                             
[XMP]           Exposure Program                : Program AE                                                                                      
[XMP]           Shutter Speed Value             : 1/20                                                                                            
[XMP]           Aperture Value                  : 2.3                                                                                                                                                                                                                                                
[XMP]           Brightness Value                : -0.95                                                                                           
[XMP]           Exposure Compensation           : 0
[XMP]           Max Aperture Value              : 2.3
[XMP]           Subject Distance                : 0.327 m
[XMP]           Metering Mode                   : Center-weighted average
[XMP]           Focal Length                    : 3.7 mm
[XMP]           Sensing Method                  : One-chip color area
[XMP]           Scene Type                      : Directly photographed
[XMP]           Custom Rendered                 : Custom
[XMP]           Exposure Mode                   : Auto
[XMP]           White Balance                   : Auto
[XMP]           Digital Zoom Ratio              : 0
[XMP]           Focal Length In 35mm Format     : 56 mm
[XMP]           Scene Capture Type              : Standard
[XMP]           Contrast                        : Normal
[XMP]           Saturation                      : Normal
[XMP]           Sharpness                       : Normal
[XMP]           Subject Distance Range          : Macro
[XMP]           GPS Version ID                  : 2.0.0.0
[XMP]           GPS Map Datum                   : WGS-84
[XMP]           GPS Altitude Ref                : Above Sea Level
[XMP]           GPS Altitude                    : 800.37 m
[XMP]           GPS Latitude Ref                : N
[XMP]           GPS Latitude                    : XXXXX
[XMP]           GPS Longitude Ref               : E
[XMP]           GPS Longitude                   : XXXXX
[XMP]           Date/Time Modified              : 2021:04:14 15:58:53
[XMP]           Image Width                     : 1728
[XMP]           Image Height                    : 2304
[XMP]           Orientation                     : Horizontal (normal)
[XMP]           Y Cb Cr Positioning             : Centered
[XMP]           X Resolution                    : 72
[XMP]           Y Resolution                    : 72
[XMP]           Resolution Unit                 : inches
[XMP]           Make                            : Xiaomi
[XMP]           Camera Model Name               : Redmi Note 9 Pro (curtana)
[XMP]           Software                        : Google
[XMP]           Metadata Date                   : 2021:04:14 15:58:53
[XMP]           Create Date                     : 2021:04:14 15:58:53
[XMP]           Modify Date                     : 2021:04:14 15:58:53
[XMP]           Date Time Original              : 2021:04:14 15:58:53
[XMP]           Date UTC                        : 2021:04:14 15:58:53
[XMP]           Modification Date               : 2021:04:14 15:58:53
[XMP]           Date Created                    : 2021:04:14 15:58:53
[XMP]           Urgency                         : 0 (reserved)
[XMP]           Categories                      : XXXXXX
[XMP]           Tags List                       : XXXXX
[XMP]           Flash Fired                     : False
[XMP]           Flash Return                    : No return detection
[XMP]           Flash Mode                      : Off
[XMP]           Flash Function                  : False
[XMP]           Flash Red Eye Mode              : False
[XMP]           Components Configuration        : Y, Cb, Cr, -
[XMP]           ISO                             : 610
[XMP]           Directory Item Mime             : image/jpeg
[XMP]           Directory Item Semantic         : Primary
[XMP]           Directory Item Length           : 0
[XMP]           Directory Item Padding          : 0
[XMP]           Directory Item Mime             : video/mp4
[XMP]           Directory Item Semantic         : MotionPhoto
[XMP]           Directory Item Length           : 3722258
[XMP]           Directory Item Padding          : 0
[XMP]           Last Keyword XMP                : XXXXX
[XMP]           Hierarchical Subject            : XXXXX
[XMP]           Catalog Sets                    : XXXXX
[XMP]           Region Name                     : XXXXX
[XMP]           Region Type                     : Face, Face, Face
[XMP]           Region Area X                   : 0.50463, 0.770255, 0.439815
[XMP]           Region Area Y                   : 0.886719, 0.435113, 0.311632
[XMP]           Region Area W                   : 0.3125, 0.34375, 0.25463
[XMP]           Region Area H                   : 0.206597, 0.343316, 0.236979
[XMP]           Region Area Unit                : normalized, normalized, normalized
[XMP]           Region Person Display Name      : XXXXX
[XMP]           Region Rectangle                : 0.34838, 0.78342, 0.3125, 0.206597, 0.59838, 0.263455, 0.34375, 0.343316, 0.3125, 0.193142, 0.25463, 0.236979
[XMP]           Subject                         : XXXXX
```

rickysarraf commented:
So the length of the file is `1164414`:

```
$ ls -l Processed/photos-backup-rmi9/PXL_20210414_155853538.MP.jpg
-rwxr-xr-x 1 rrs rrs 1164414 Jul 31 00:09 Processed/photos-backup-rmi9/PXL_20210414_155853538.MP.jpg*
```


while the MP length reported is: `3722258`, which makes the non-MP data at `-2557844`. So the XMP headers are broken somehow. :-(
And that's what makes it all crash.

Do you think it makes sense to log all this as error and still not crash ? But the problem I see is that the logged error will be lost in the huge pile. No users will notice that they may have some broken files.

rickysarraf commented:
> Do you think it makes sense to log all this as error and still not crash ? But the problem I see is that the logged error will be lost in the huge pile. No users will notice that they may have some broken files.

But I think it is right to report with a crash in this application's use case. Otherwise nothing will ever get fixed. :-)

Meanwhile, I tracked the same photo on GPhotos to see how it deals with it.  The reported size from GPhoto is consistent with the file system size I reported above. GPhoto reports it at `1.1 MiB`. Also, the motion video, as one would expect in this case, is very very short rendered in GPhotos.

kvalev commented:
Thanks for the report! I should have assumed that the offset in the other Google motion photo format can be invalid. I will add a fix - it should recover exactly the same way as the other metadata.

This way, incorrect metadata in the Google motion photos should not lead to a crash - instead in case the metadata was wrong and the recovery process failed (meaning no valid mp4 file was found in the image), an error will now be logged.

kvalev committed:
MP: Additional fallbacks in case of erroneous offset metadata

kvalev committed:
MP: Unit tests

kvalev commented:
The latest commits should cover few more use cases where the metadata might be incorrect. In any case the implementation will try to find the embedded mp4 and if it doesn't, it will log an error. Can you try it once more and let me know if you encounter any other issues or unexpected behavior? It seems I dont have any faulty motion photos, as I never had those issues when testing :D

Fore debugging purposes it might be helpful to let it crash, but I am a bit reluctant to do so, even if this means that some errors might be overlooked in verbose logs. Another option I am considering is to let it crash only when the user has enabled the experimental features.

rickysarraf commented:
Thanks @kvalev 

Your change seems to be working perfect now. The scanning is on for 7+ hours and I haven't seen any crash at all.

I have noticed an odd behavior though. Please consider the below image, where you'll see on the LHS, the generated thumbnails, of the MP pictures from the RHS. I understand that generated thumbnails reside in the sidecar folder and shouldn't be part of the main page view in PhotoPrism.

My guess is that for MP pictures, the video is extracted and then PhotoPrism [mis]treats it as a new unique video file and ends up generating a thumbnail for it. Oh!! And all these generated thumbnails are put under **Review**, which I understand is the right thing being done by PhotoPrism, given the size/quality of these thumbnail images.


![image](https://user-images.githubusercontent.com/65480/134804671-9b016607-aadd-4852-9e2c-f14e7c1e13b0.png)

kvalev commented:
@rickysarraf Thanks for the info! I will be able to take a look at it towards the end of the week.

rickysarraf commented:
> @rickysarraf Thanks for the info! I will be able to take a look at it towards the end of the week.

Today morning, magically, all those annoying thumbnails were gone.

After taking a closer look at the logs, I noticed that I have a scheduled job to run `cleanup` and `purge` once a day. That seems to have done the magic.

```
Sep 30 13:08:26 lenovo docker[519905]: time="2021-09-30T07:38:26Z" level=info msg="cleanup: updating photo counts"
Sep 30 13:08:28 lenovo docker[519905]: time="2021-09-30T07:38:28Z" level=info msg="cleanup: updating preview images"
Sep 30 13:08:34 lenovo docker[519905]: time="2021-09-30T07:38:34Z" level=info msg="cleanup: removed 502 index entries and 6071 orphan thumbnails in 3m11.642196034s"
Sep 30 13:08:34 lenovo docker[519905]: time="2021-09-30T07:38:34Z" level=info msg="closed database connection"
Sep 30 13:08:34 lenovo systemd[6958]: photoprism-cleanup.service: Succeeded.
```

I am still not sure why those thumbnails had shown up in the main view.

And Michael has committed a fix to that behavior: 0d772c4dc3864b0dd247deb7765cb4fdc4c7789e
So I guess there's nothing wrong in this PR.

kvalev committed:
MP: Remove debug log messages

kvalev committed:
MP: Dedicated event to be shown in the UI

The indexing UI will now show the correct message when motion photos are being
indexed and the video portion is being extracted.

kvalev committed:
MP: Extend the reset command to remove transcoded videos and motion photo videos

kvalev commented:
I now had a bit more time to look into it and I could not reproduce it. It must have been an upstream issue - I reset my test instance multiple times and didnt experience it once, which I guess is a good sign.

kvalev committed:
Merge branch 'preview' into kvv-support-motion-photos

kvalev commented:
The feature has been released to `kvalev/photoprism:211006-2e5220a`.
